File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.Stack;

public class Solution {

    // Method to check if the string contains valid parentheses
    public boolean isValid(String s) {
        // Use a stack to keep track of opening brackets
        Stack<Character> stack = new Stack<>();
        
        // Iterate through each character in the string
        for (char c : s.toCharArray()) {
            // If the character is an opening bracket, push it onto the stack
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } 
            // If the character is a closing bracket
            else {
                // If the stack is empty or the top of the stack does not match the closing bracket, return false
                if (stack.isEmpty() || 
                   (c == ')' && stack.pop() != '(') ||
                   (c == '}' && stack.pop() != '{') ||
                   (c == ']' && stack.pop() != '[')) {
                    return false;
                }
            }
        }
        
        // If the stack is empty, all brackets were properly closed; otherwise, return false
        return stack.isEmpty();
    }

    // Main method for testing the functionality
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test cases
        String[] testCases = { "()", "()[]{}", "(]", "([])", "([)]", "{[]}", "" };
        
        // Run each test case
        for (String testCase : testCases) {
            System.out.println("Input: " + testCase);
            System.out.println("Output: " + solution.isValid(testCase));
            System.out.println();
        }
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s):
        # Use a stack to keep track of opening brackets
        stack = []
        # Map to match opening and closing brackets
        bracket_map = {')': '(', '}': '{', ']': '['}
        
        # Iterate through each character in the string
        for char in s:
            # If the character is a closing bracket
            if char in bracket_map:
                # Pop the top of the stack if it's not empty, otherwise use a dummy value
                top_element = stack.pop() if stack else '#'
                # If the top of the stack doesn't match the corresponding opening bracket, return False
                if bracket_map[char] != top_element:
                    return False
            else:
                # Push opening brackets onto the stack
                stack.append(char)
        
        # If the stack is empty, all brackets were properly closed; otherwise, return False
        return not stack


# Testing the Solution
if __name__ == "__main__":
    solution = Solution()
    
    # Test cases
    test_cases = ["()", "()[]{}", "(]", "([])", "([)]", "{[]}", ""]
    
    # Run each test case
    for test in test_cases:
        print("Input: {}".format(test))
        print("Output: {}".format(solution.isValid(test)))
        print()
